core/network/src/sync.rs

```diff
@@ -50,6 +54,147 @@ enum PeerSyncState<B: BlockT> {
 	DownloadingStale(B::Hash),
 }
 
+type PendingJustification<B> = (<B as BlockT>::Hash, NumberFor<B>);
+
+/// Manages pending block justification requests.
+struct PendingJustifications<B: BlockT> {
+	justifications: HashSet<PendingJustification<B>>,
+	pending_requests: VecDeque<PendingJustification<B>>,
+	peer_requests: HashMap<NodeIndex, PendingJustification<B>>,
+	previous_requests: HashMap<PendingJustification<B>, Vec<(NodeIndex, Instant)>>,
+}
+
+impl<B: BlockT> PendingJustifications<B> {
+	fn new() -> PendingJustifications<B> {
+		PendingJustifications {
+			justifications: HashSet::new(),
+			pending_requests: VecDeque::new(),
+			peer_requests: HashMap::new(),
+			previous_requests: HashMap::new(),
+		}
+	}
+
+	fn dispatch(&mut self, peers: &HashMap<NodeIndex, PeerSync<B>>, protocol: &mut Context<B>) {
+		// clean up previous failed requests so we can retry again
+		for (_, requests) in self.previous_requests.iter_mut() {
+			requests.retain(|(_, instant)| instant.elapsed() < JUSTIFICATION_RETRY_WAIT);
+		}
+
+		let mut peers = peers.iter().filter_map(|(peer, status)| {
```

Comment:
Probably early exit if `pending_requests.is_empty()` to avoid extra iterations + allocations?

Reply:
it can be done by collecting into a `Result<VecDeque<_>, ()>` instead.

Reply:
(this is resolved now)